rfcs/require-by-string.md

```diff
@@ -0,0 +1,310 @@
+
+To solve this issue, we propose transforming every path that is passed to `require` into an equivalent absolute path and using this to cache, regardless of whether it was passed in as a relative, absolute, or aliased path. This way, a module's return value is stored in both a unique and consistent way across different files.
+
+#### DataModel as VFS
+
+In the Roblox engine, the DataModel will act as a virtual file system. At the root of this VFS is the DataModel itself. This will allow packages from Luau to be imported and exported freely without needing to consider the platform they are being used on.
+
+All paths used in the Roblox engine must refer to a location in the DataModel. They cannot be used to access files on disk.
+
+```lua
+-- MyModule location: game:GetService("ReplicatedStorage").MyModule
+
+-- Require by absolute path
+local MyModule = require("/ReplicatedStorage/MyModule")
+
+-- Require by relative path
+-- (From game:GetService("ReplicatedStorage").MyOtherModule)
+local MyModule = require("MyModule")
+```
+
+##### Considerations
+
+Within the Roblox engine, we will have to handle replication and waiting for modules to be loaded into the DataModel before requiring them. However, this is outside of the scope of this RFC and will be discussed internally.
+
+#### Platform compatibility
+
+For compatibility across platforms, we will automatically map `/` onto `\`.
+
+## Drawbacks
+
+### Backwards compatibility
+
+Luau libraries are already not compatible with existing Lua libraries. This is because Lua favors the `.` based require syntax instead and relies on the `LUA_PATH` environment variable to search for modules, whereas Luau currently supports a basic require-by-string syntax.
+
+- Libraries are fully compatible with the Roblox engine, as require-by-string is currently unimplemented.
+- Luau currently implements relative paths in relation to the current working directory. 
+- We propose changing this behavior and breaking backwards compatibility on this front.
+- With the current implementation, requiring a library that itself contains relative-path require statements [can become a mess](https://github.com/Roblox/luau/issues/959) if the Luau VM is not launched from the "correct" working directory.
+- We propose the following change: relative paths passed to require statements will be evaluated in relation to the requiring file's location, not in relation to the current working directory.
+    - Caveat: if the relative-path require is called directly in the Luau CLI, this will remain relative to the current working directory, as there is no sense of a "requiring file" here.
+- If this causes issues, we can later introduce a default alias for the current working directory (e.g. `@CWD`).
+
```

Comment:
I am a bit confused about aliases. From what I understood by reading this RFC, they are a shorthand for paths where code loaded by require is stored. It feels a bit heavyweight for users to maintain a .luarc.

Some languages (such as R) have the concept of a library which is essentially a directory of packages.
This library is the search path used for loading packages. When you say `require(Package)`, the system looks for a directory called `Package` in the library directory (set to a sensible default value based on user's OS). The user can add multiple paths to the `library` and they are searched in order until `Package` is found. 
The library search path (a vector datatype) in our case could be managed by Roblox studio with user accessible APIs to add, remove, and reorder search paths. This could save users from maintaining a .luarc with aliases.

Reply:
Adding a comment to summarize an internal conversation about this: other languages typically implement the behavior that @aviralg describes using environment variables that define where to look for packages (which might also answer @BradSharp's question):

- Python: `PYTHONPATH`
- Lua (not Luau): `LUA_PATH` / `LUA_CPATH`
- R: `R_LIBS` / `R_LIBS_USER` / `R_LIBS_SITE`

Within Roblox Studio, providing this flexibility through some global variable (e.g. `LUAU_PATH`) doesn't seem like an ideal approach: requiring certain modules may not work for some users, depending on how they've configured their `LUAU_PATH`. It probably makes more sense to just fix a location for "requirable" libraries/packages in Studio.

Outside of Roblox (Luau CLI), it might make sense to add `LUAU_PATH` in the future. This won't be covered in this RFC but is definitely compatible with the changes proposed here: for example, `LUAU_PATH` could always include the "current" file's directory (either explicitly or implicitly), ensuring that relative paths continue to work in the future.

Reply:
It feels like a fairly straightforward extension to this proposal would say that, absent an alias definition in .luaurc, Luau CLI will attempt to find a directory with the same name as the alias using LUAU_PATH envvar or something like `-Lpath` command line argument. Does that make sense? Want to make sure this is the same as above - we wouldn't say that `require("Foo/bar")` tries to find `"Foo"` in a different path, but we would say that `require("@Foo/bar")` just works? Or is the idea here that we would keep alias explicit only, and instead support `require("Foo")`?

Reply:
Regardless I feel like we should specify some handling of this problem, whether through aliases or through unaliased paths, in the RFC, because without it I feel like we would be breaking some existing scripts (eg our benchmark scripts all require bench_support, how will that need to change?) without solid guidance as to what to use as a replacement. IOW because this RFC is changing existing CLI behavior I think we should make sure we specify a reasonably final version of the CLI behavior that replaces it in one go, instead of in two steps.

Reply:
I think this would also address your comment here, @zeux: https://github.com/Roblox/luau/pull/969#issuecomment-1648202266

Taking this approach would mean implementing support for `LUAU_PATH`, which would serve the same purpose as `LUA_PATH` in Lua. We wouldn't necessarily need to force relative paths to begin with `./` if we keep aliases explicit (with `@`). By default, a given path would be treated as relative to the requiring file, and we would fallback to using `LUAU_PATH`.

Aliases would be added functionality on top of all of this (which, up to this point is really similar to Lua). Any path beginning with `@` would be resolved using alias maps that are defined in `.luaurc` files.